functions/diff.js

```diff
@@ -3,36 +3,52 @@ const parse = require('./parse')
 const diff = (version1, version2) => {
   const v1 = parse(version1)
   const v2 = parse(version2)
-  if (v1.compare(v2) === 0) {
+  const comparison = v1.compare(v2)
+
+  if (comparison === 0) {
     return null
-  } else {
-    const hasPre = v1.prerelease.length || v2.prerelease.length
-    const prefix = hasPre ? 'pre' : ''
-    const defaultResult = hasPre ? 'prerelease' : ''
-
-    if (v1.major !== v2.major) {
-      return prefix + 'major'
-    }
-    if (v1.minor !== v2.minor) {
-      return prefix + 'minor'
-    }
-
-    if (v1.patch !== v2.patch) {
-      return prefix + 'patch'
-    }
-
-    if (!v1.prerelease.length || !v2.prerelease.length) {
-      if (v1.patch) {
-        return 'patch'
-      }
-      if (v1.minor) {
-        return 'minor'
-      }
-      if (v1.major) {
-        return 'major'
-      }
-    }
-    return defaultResult // may be undefined
   }
+
+  const v1Higher = comparison > 0
+  const highVersion = v1Higher ? v1 : v2
+  const lowVersion = v1Higher ? v2 : v1
+  const highHasPre = !!highVersion.prerelease.length
+
+  // add the `pre` prefix if we are going to a prerelease version
+  const prefix = highHasPre ? 'pre' : ''
+
+  if (v1.major !== v2.major) {
+    return prefix + 'major'
+  }
+
+  if (v1.minor !== v2.minor) {
+    return prefix + 'minor'
+  }
+
+  if (v1.patch !== v2.patch) {
+    return prefix + 'patch'
+  }
+
+  // at this point we know stable versions match but overall versions are not equal,
+  // so either they are both prereleases, or the lower version is a prerelease
+
+  if (highHasPre) {
+    // high and low are preleases
+    return 'prerelease'
+  }
+
+  if (lowVersion.patch) {
+    // anything higher than a patch bump would result in the wrong version
+    return 'patch'
+  }
+
+  if (lowVersion.minor) {
+    // anything higher than a minor bump would result in the wrong version
+    return 'minor'
+  }
+
+  // bumping major/minor/patch all have same result
+  return 'major'
```

Comment:
Related to the last part of the PR description, this would make more sense to me if this was all replaced with `return 'patch'`

The way `npm version` behaves with major and minor bumps when the current version is a prerelease and parts are `0` is a bit confusing to me

Specifically I don't understand why there is a 0 special case [here](https://github.com/npm/node-semver/blob/27817677794f592b592bf6181a80a4824ff762b2/classes/semver.js#L231) and [here](https://github.com/npm/node-semver/blob/27817677794f592b592bf6181a80a4824ff762b2/classes/semver.js#L215-L219)

Reply:
[special case for inc minor](https://github.com/npm/node-semver/blob/27817677794f592b592bf6181a80a4824ff762b2/classes/semver.js#L231)

The comments tell the story there:
>        // If this is a pre-minor version, bump up to the same minor version.
>        // Otherwise increment minor.

A preminor version is a version whose patch is `0` and has a prerelease identifier.  Bumping a minor from that is a matter of dropping the prerelease identifier.

[special case for inc major
](https://github.com/npm/node-semver/blob/27817677794f592b592bf6181a80a4824ff762b2/classes/semver.js#L215-L219)

Same as before, only we are looking at minor and patch.  A premajor version is a version whose minor and patch are `0` and has a prerelease identifier.  Bumping a major from that is a matter of dropping the prerelease identifier.

Reply:
I guess it's a bit strange that you can go from `1.0.1-0` to `1.1.0` with `npm version minor` but you can't go from `1.0.0-0` to `1.1.0` because the command works differently for that case.

I still don't really get why 0 is special 🤷. Seems a bit inconsistent

Reply:
It does appear inconsistent but that is due to the nature of what `1.0.0-0` is. It's the prerelease for a semver major version.  Dropping that prerelease is a major, there is no smaller increment you can make outside of increasing the prerelase identifier.

Reply:
Right yes that makes sense. The thing that's confusing I think is that 0.1.0 to 1.0.0 is concretely major, but I'm not sure going from 1.0.0-0 to 1.0.0 could be classed as a major change, it's just going prerelease to stable. Maybe patch/minor/major doesn't really make sense as a thing when going prerelease to stable 🤷

I think the difference between 1.0.1-0 and 1.0.1 has the same weight

Reply:
It does make sense if the result from diff when we're going from prerelease to stable release is meant to represent the diff between the previous stable version and the new one 💡, but that's a step I wouldn't expect the diff to be doing internally